criu/files-reg.c

```diff
@@ -1380,12 +1388,318 @@ static bool should_check_size(int flags)
 	return true;
 }
 
+/* Gets the build-id (If it exists) from 32-bit ELF files.
+ * Returns the number of bytes of the build-id if it could
+ * be obtained, else -1.
+ */
+static int get_build_id_32(Elf32_Ehdr *file_header, unsigned char **build_id,
+				const int fd, size_t mapped_size)
+{
+	int size, num_iterations;
+	size_t file_header_end;
+	Elf32_Phdr *program_header, *program_header_end;
+	Elf32_Nhdr *note_header, *note_header_end;
+
+	file_header_end = (size_t) file_header + mapped_size;
+	if (file_header >= (Elf32_Ehdr *) (file_header_end - sizeof(Elf32_Ehdr)))
+		goto err;
+
+	/* 
+	 * If the file doesn't have atleast 1 program header entry, it definitely can't
+	 * have a build-id.
+	 */
+	if (!file_header->e_phnum) {
+		pr_warn("Couldn't find any program headers for file with fd %d\n", fd);
+		goto err;
+	}
+
+	program_header = (Elf32_Phdr *) (file_header->e_phoff + (char *) file_header);
+	if (program_header <= (Elf32_Phdr *) file_header)
+		goto err;
+
+	program_header_end = (Elf32_Phdr *) (file_header_end - sizeof(Elf32_Phdr));
+	num_iterations = file_header->e_phnum + 1;
+
+	/* 
+	 * If the file has a build-id, it will be in the PT_NOTE program header 
+	 * entry AKA the note sections.
+	 */
+	while (num_iterations-- && program_header <= program_header_end &&
+			program_header->p_type != PT_NOTE)
```

Comment:
I don't insist here, but I'd just call it `i` insted:
```
for (i = 0; i < file_header->e_phnum && program_header < program_header_end; i++) {
     if (program_header->p_type == PT_NOTE)
          break;
}
```
Also, note that `program_header <= program_header_end` should be without `=`.

Reply:
Wouldn't this be a bit harder to read? I find it easier to look at the entire break condition together like in the while loop.